clustered/server/src/main/java/org/ehcache/clustered/server/store/ClusterTierActiveEntity.java

```diff
@@ -586,40 +603,129 @@ public void synchronizeKeyToPassive(PassiveSynchronizationChannel<EhcacheEntityM
           }
-        });
-      if (!mappingsToSend.get().isEmpty()) {
-        syncChannel.synchronizeToPassive(new EhcacheDataSyncMessage(mappingsToSend.get()));
-        mappingsToSend.set(new HashMap<>());
-        size.set(0);
+        }
+
+        if (interrupted) {
+          // get is finite memory operation..give it time to complete even if async get task is not interrupted
+          // for some reason so that we can get the actual cause of exception.
+          asyncGet.get(10, TimeUnit.SECONDS);
+          // the get threads actual execution status precedes our own interruption status
+          throw new InterruptedException();
+        } else {
+          asyncGet.get();
+        }
+      } catch (InterruptedException | ExecutionException | TimeoutException e) {
+        throw new RuntimeException(e);
       }
     }
     sendMessageTrackerReplication(syncChannel, concurrencyKey - 1);
 
     LOGGER.info("Sync complete for concurrency key {}.", concurrencyKey);
   }
 
+  private void doGetsForSync(int segmentId, BlockingQueue<EhcacheDataSyncMessage> messageQ, Thread waitingThread) {
+    ServerSideServerStore store = stateService.getStore(storeIdentifier);
+    int numKeyGets = 0;
+    long dataSize = 0;
+    boolean success = false;
+    try {
+      Set<Long> keys = store.getSegmentKeySets().get(segmentId);
+      int remainingKeys = keys.size();
+      Map<Long, Chain> mappingsToSend = new HashMap<>(computeInitialMapCapacity(remainingKeys));
+      boolean capacityAdjusted = false;
+      for (Long key : keys) {
+        final Chain chain;
+        try {
+          chain = store.get(key);
+          numKeyGets++;
+        } catch (TimeoutException e) {
+          throw new AssertionError("Server side store is not expected to throw timeout exception");
+        }
+        for (Element element : chain) {
+          dataSize += element.getPayload().remaining();
+        }
+        mappingsToSend.put(key, chain);
+        if (dataSize > dataSizeThreshold || numKeyGets >= dataGetsThreshold) {
+          messageQ.put(new EhcacheDataSyncMessage(mappingsToSend));
+          if (!capacityAdjusted && segmentId == 0) {
+            capacityAdjusted = true;
+            adjustInitialCapacity(numKeyGets);
+          }
+          remainingKeys -= numKeyGets;
+          mappingsToSend = new HashMap<>(computeMapCapacity(remainingKeys, numKeyGets));
+          dataSize = 0;
+          numKeyGets = 0;
+        }
+      }
+      if (!mappingsToSend.isEmpty()) {
+        messageQ.put(new EhcacheDataSyncMessage(mappingsToSend));
+      }
+      messageQ.put(SYNC_MESSAGE_COMPLETION_MARKER);
+      success = true;
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    } finally {
+      if (!success) {
+        // ensure waiting peer thread is interrupted, if we run into trouble
+        waitingThread.interrupt();
+      }
+    }
+  }
+
+  private int computeMapCapacity(int remainingSize, int expectedGets) {
```

Comment:
Can you comment on why you did this split or sizing?

Reply:
@henri-tremblay Not sure I fully understand the question..
    
   Anyway, the logic here is simple: 
               remainingSize is the number of keys left in the chainmap yet to be send
               expectedGets is the max number of items that could be put in a single hashmap
both varies depending on the size of the chains and number of keys. For e.g I may start with an initial size of configured max gets and then soon discover that the size of 2MB is reached for this particular cache before it even reaches 75% of max gets. This will automatically adjust the hashmap capacity for the next iteration. I also re-update the initial size so that for the next concurrency, the initial capacity matches the type of cache  (e.g more data and less keys) we have. 

Ofcourse the logic assumes some sort of symmetry at the concurrency level,, but it is possible that some keys has chain with LARGE data and some keys has small. But on a larger sweep that should even out.. even if it is not, this should perform better as the sizes are reset back to a minimum of 32 and not 16 when a cache is large and when a cache is very small it starts with initial size of 16.

Reply:
Can you write that in a comment in the code? :-)

Reply:
sure.. I will put a method header for the compute* methods..